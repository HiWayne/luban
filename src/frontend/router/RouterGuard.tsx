import { FC, useEffect, useState } from 'react';
import { Navigate, useParams } from 'react-router-dom';
import { AccessDenied, Loading } from 'components/index';

interface Auth {
  loading: boolean;
  auth: boolean;
  login: boolean;
}

interface RouterGuardProps {
  from: string;
  element: FC;
  permissions?: string[];
  getAuth?: (permissions?: string[]) => Promise<Auth>;
  __test__?: { error: boolean; data: any };
}

const RouterGuard: FC<RouterGuardProps> = ({
  from,
  element: Element,
  permissions,
  getAuth,
}) => {
  const [{ loading, auth, login }, setStatus] = useState(
    typeof getAuth === 'function'
      ? {
          loading: false,
          auth: false,
          login: false,
        }
      : {
          loading: false,
          auth: true,
          login: true,
        },
  );

  const params = useParams() || {};

  useEffect(() => {
    (async () => {
      if (typeof getAuth === 'function') {
        setStatus((status) => ({
          ...status,
          loading: true,
        }));
        const response = await getAuth(permissions);
        setStatus(response);
      }
    })();
  }, []);

  if (loading) {
    return <Loading size="large" />;
  }
  if (auth) {
    return <Element />;
  } else if (!login) {
    return <Navigate to="/login" state={{ ...params, __from: from }} replace />;
  } else {
    return <AccessDenied />;
  }
};

export default RouterGuard;
