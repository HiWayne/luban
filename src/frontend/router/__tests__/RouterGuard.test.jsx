import { render } from '@testing-library/react';
import { Route, Routes, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import RouterGuard from '../RouterGuard';

test('RouterGuard 测试 登录无权限', () => {
  const history = createMemoryHistory();
  const { asFragment } = render(
    <Router location={history.location} navigator={history}>
      <Routes>
        <Route path="/login" element={<div>login</div>} />
        <Route
          path="/"
          element={
            <RouterGuard
              from="/"
              element={() => <div />}
              permissions={['normal_manager']}
              __test__={{
                error: false,
                data: {
                  status: 1,
                  data: {
                    identity: [],
                  },
                },
              }}
            />
          }
        />
      </Routes>
    </Router>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('RouterGuard 测试 未登录', () => {
  const history = createMemoryHistory();
  const { asFragment } = render(
    <Router location={history.location} navigator={history}>
      <Routes>
        <Route path="/login" element={<div>login</div>} />
        <Route
          path="/"
          element={
            <RouterGuard
              from="/"
              element={() => <div />}
              permissions={['normal_manager']}
              __test__={{
                error: false,
                data: {
                  status: 2,
                  data: null,
                },
              }}
            />
          }
        />
      </Routes>
    </Router>,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('RouterGuard 测试 有权限', () => {
  const history = createMemoryHistory();
  const { asFragment } = render(
    <Router location={history.location} navigator={history}>
      <Routes>
        <Route path="/login" element={<div>login</div>} />
        <Route
          path="/"
          element={
            <RouterGuard
              from="/"
              element={() => <div />}
              permissions={['normal_manager']}
              __test__={{
                error: false,
                data: {
                  status: 1,
                  data: { identity: ['normal_manager'] },
                },
              }}
            />
          }
        />
      </Routes>
    </Router>,
  );
  expect(asFragment()).toMatchSnapshot();
});
