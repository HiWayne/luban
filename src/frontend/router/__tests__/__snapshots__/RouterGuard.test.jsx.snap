// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterGuard 测试 有权限 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`RouterGuard 测试 未登录 1`] = `<DocumentFragment />`;

exports[`RouterGuard 测试 登录无权限 1`] = `
<DocumentFragment>
  <div>
    没有权限
  </div>
</DocumentFragment>
`;
